nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+maintaining a handler side persistent mapping between caller and handler identifiers, the handler may choose to "leak"
+this detail to the caller, which would need to concatinate multiple identifiers for a given call. The example above
+becomes:
+
+```js
+Start(
+  'charge',
+  customerId + '/' + transactionId,
+  { amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
+);
+Cancel('charge', customerId + '/' + transactionId);
+```
+
+> NOTE: This looks a lot like a nested REST style URL, e.g: `/customers/{customerId}/charges/{transactionId}`.
+
+> To take this approach to the extreme, `operation` and `id` could be combined into a single `path` argument, as shown
+> below:
+>
+> - `Start(path, input, callbackURL)`
+> - `Cancel(path)`
+> - `GetResult(path)`
+> - `Describe(path)`
+
+##### Mapping from client to handler identifiers
+
+To avoid leaking the identifier construction logic to the client, the handler would need a mapping of client identifiers
+to handler identifiers. Temporal can provide a strongly consistent store out of the box for this mapping.
+
+A single execution would need to map to multiple client identifiers to account for workflow starts and updates.
+
+Looking up an identifier in persistent storage added latency, and should be opted in to.
+
+### Handler generated identifiers
+
+With this approach, the handler returns a reference Id in response to a start call, which the caller should persist for
+use with the pre-defined RPCs as shown below:
+
+- `Start(operation, input, callbackURL, cancelBeforeStarted)` - handler responds inline or returns a reference Id
+  ([`*`](#cancel-before-started))
+- `Cancel(reference_id)` - cancel operation by handler provided reference Id
+- `GetResult(reference_id)` - get operation result by handler provided reference Id
+- `Describe(reference_id)` - get operation status by handler provided reference Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+The handler _should_ either sign or qualify caller-provided reference Ids when dealing with untrusted sources and
+multi-tenant applications.
+
+In order to support stateless clients, the handler may provide an additional set of APIs to resolve client input to a
+reference Id. This may be something that is built into the protocol or implemented ad-hoc. To support this at the
+protocol level, we would need to add the following APIs:
+
+- `Resolve(operation, input) -> reference_id`
+
+-- OR --
+
+- `Cancel(operation, input)` - cancel operation by handler derived Id
+- `GetResult(operation, input)` - get operation result by handler derived Id
+- `Describe(operation, input)` - get operation status by handler derived Id
+
+NOTE: This looks a lot like the APIs for the "client generated Ids" approach, except the ID is structured instead of a
+single string.
+
+#### Cancel before started
```

Comment:
I don't think this should be supported in Nexus. I think an operation does not exist until `Start` returns and _no_ calls can be made on an operation that does not exist yet. If the Temporal workflow side of Nexus wants to implement cancel-before-start by never sending the start, so be it. But a cancel after start should always wait on start to return before issuing the cancel.

Reply:
+1

Reply:
You're forcing the caller to issue an additional call and delaying the delivery of the cancellation. We can discuss how this would be implemented with workflows.

Reply:
I just want to make sure from a Nexus perspective, Nexus servers do not have to accept cancellation from operations that have not returned from start.

Reply:
Would rather handle that as a Temporal thing in the server than say it's optional in Nexus. I assume the server has similar logic for child workflows and activity workflows when it receives a cancel for a not-started/asynchronously-waiting-to-start thing.

Reply:
> Would rather handle that as a Temporal thing in the server than say it's optional in Nexus

How would Temporal be able to achieve this without it being a part of the Nexus protocol?

> I assume the server has similar logic

Not sure I understand what you mean here.

The server (history) would be the one issuing these cancel-maybe-started requests, if the handler is backed by Temporal, it would be trivial to implement for most operations.